Support for installing RHDH using Operator (#62)
redhat-developer/red-hat-developers-documentation-rhdh: installation/proc-install-rhdh-operator.adoc

[id='proc-install-rhdh-operator_{context}']
= Installing {product} using the Operator

[IMPORTANT]
====
The {product} Operator is a Technology Preview feature only.

Technology Preview features are not supported with Red Hat production service level agreements (SLAs), might not be functionally complete, and Red Hat does not recommend using them for production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information on Red Hat Technology Preview features, see https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Scope].
====

This section describes the installation procedure for installing {product} on your OpenShift Container Platform using the Operator.

//You can use the {product} Operator in Red Hat OpenShift (OpenShift) to install {product}.

.Prerequisites

> * You are logged in as an administrator the OpenShift web console.

Reviewer: ```suggestion
* You are logged in as an administrator on the OpenShift web console.
```

* You have configured the appropriate roles and permissions within your project to create an application. See the link:https://docs.openshift.com/container-platform/4.14/applications/index.html[Red Hat OpenShift documentation on Building applications] for more details.

> * If installing a CI build from quay.io/rhdh, you have added a link:https://gist.github.com/nickboldt/d54ba9d7264a5480d14eea6c3bd2ecdf[RHDH Operator Catalog Source] and a link:https://gist.github.com/nickboldt/84bea28da11ec5aab549c182818356e0#file-oc-pull-secret-update-project-sh[pull secret].

Reviewer: Can we be more explicit and add the exact commands that users need to run?
Essentially:
- how to `oc login`
- commands to download and execute the scripts

Wondering if a dedicated section about installing CI builds wouldn't make more sense then..
WDYT?

Author: Would Installing CI Builds be specific to the Operator Install before RHDH-1.1, I presume after RHDH-1.1 is released, we would not need this for the Operator Install?

If it's only related to the current situation before RHDH-1.1, and not relevant thereafter, it might be better to just and add the exact commands that users need to run in this section.

Reviewer: Yes, for now, this is specific to the pre-1.1 version, and should not appear in the final 1.1 docs, I think.
Sounds good then, as long as we add those commands, so that the user knows exactly what to run.

Reviewer: I just tried both scripts on a fresh cluster, and noticed a few things:
- adding the pull secret using https://gist.github.com/nickboldt/84bea28da11ec5aab549c182818356e0#file-oc-pull-secret-update-project-sh does not seem to be sufficient. The pull secret is created in a `mybackstage` namespace, but the operator itself could not be installed in the `openshift-operators` namespace because it could not pull the `rhdh-rhel9-operator` image from Quay. This was fixed by adding my Quay token to the cluster global pull secret
- and by the way, the script surprisingly [creates a new project](https://gist.github.com/nickboldt/84bea28da11ec5aab549c182818356e0#file-oc-pull-secret-update-project-sh-L15-L17), which may seem surprising since we are talking about adding a pull secret. It should be up to the user to use their own namespace, I think.

So I'd suggest the following instructions:
```suggestion
* If installing a CI build from quay.io/rhdh, make sure that:
** you have been granted permission to pull images from the `rhdh` organization on quay.io, and 
** you have added your Quay token to the cluster global pull secret (link:https://docs.openshift.com/container-platform/4.14/openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets[link]):
{code}
oc get secret/pull-secret -n openshift-config --template='{{index .data ".dockerconfigjson" | base64decode}}' > /tmp/my-global-pull-secret.yaml

oc registry login --registry="quay.io" --auth-basic="<user>:<token>" --to=/tmp/my-global-pull-secret.yaml

oc set data secret/pull-secret -n openshift-config --from-file=.dockerconfigjson=/tmp/my-global-pull-secret.yaml

rm -f /tmp/my-global-pull-secret.yaml
{code}
** you have added a link:https://gist.github.com/nickboldt/d54ba9d7264a5480d14eea6c3bd2ecdf[RHDH Operator Catalog Source]:
{code}
curl -L https://gist.githubusercontent.com/nickboldt/d54ba9d7264a5480d14eea6c3bd2ecdf/raw/76a7f166a17f86ae8e766241b9d648c9821898bb/install-rhdh-next-build.sh | bash
{code}
```

cc @nickboldt

.Procedure

> . In the *Administrator* view of the OpenShift web console, go to *Operators* → *OperatorHub* and search for Red Hat Developer Hub and install the Red Hat Developer Hub Operator. For more info, see link:https://docs.openshift.com/container-platform/4.14/operators/admin/olm-adding-operators-to-cluster.html#olm-installing-from-operatorhub-using-web-console_olm-adding-operators-to-a-cluster[Installing from OperatorHub using the web console].

Reviewer: ```suggestion
. In the *Administrator* view of the OpenShift web console, go to *Operators* → *OperatorHub*, search for Red Hat Developer Hub, and install the Red Hat Developer Hub Operator. For more info, see link:https://docs.openshift.com/container-platform/4.14/operators/admin/olm-adding-operators-to-cluster.html#olm-installing-from-operatorhub-using-web-console_olm-adding-operators-to-a-cluster[Installing from OperatorHub using the web console].
```

> . In the *Administrator* view of the OpenShift web console, go to *Operators* → *OperatorHub*, search for Red Hat Developer Hub, and install the Red Hat Developer Hub Operator. For more info, see link:https://docs.openshift.com/container-platform/4.14/operators/admin/olm-adding-operators-to-cluster.html#olm-installing-from-operatorhub-using-web-console_olm-adding-operators-to-a-cluster[Installing from OperatorHub using the web console].

Reviewer 2: Administrator view => Administrator perspective
Also instead of using a reference link for operator installation, it might be more helpful  for users if we include the operator installation steps here directly with screenshots.

. Create an OpenShift project to be used by your Backstage instance. For more information about creating a project in OpenShift, see the link:https://docs.openshift.com/container-platform/4.14/applications/projects/working-with-projects.html#creating-a-project-using-the-web-console_projects[Red Hat OpenShift documentation].

> . Switch to *Developer* mode on your Red Hat OpenShift web console.

Reviewer 2: "Developer mode" => "Developer perspective"

As contrast to helm chart installation which can only be done in the developer perspective, you do not have to switch to the Developer perspective to deploy custom resource instances for operators.  Instead, it is normally done in the Administrator perspective (as you are already there) to create custom resource instances.

Reviewer 3: true, but if we're splitting the doc in to admin steps and developer steps (with prereq on admin steps)... then it might make sense to have people always switch to the developer perspective.

[id='proc-install-rhdh-operator_{context}']

> = Installing {product} using the Operator

Reviewer 3: I notice we have helm chart installation instructions in BOTH the admin guide and getting started guide (because it can be done by an admin or a user); however this new doc for installing via operator is ONLY in the getting started guide, which is wrong since you need to be cluster admin to install via operator. 

ONCE the operator is installed, THEN a user can enable it via the `Switch to *Developer* mode on your Red Hat OpenShift web console` step and beyond.

So we need instuctions for enabling the operator via admin guide, then doc for enabling it as a user in the getting starting guide, with a link back to the admin guide for the prereq steps).

Author: @nickboldt Should we also include the full procedure for enabling the operator in the Admin guide i.e. as we have them in the User guide (without the admin steps)? Please see latest previews of both guides.

Reviewer 3: seems redundant -- why not just cross-link the sections?

For more information on Red Hat Technology Preview features, see https://access.redhat.com/support/offerings/techpreview/[Technology Preview Features Scope].
====

This section describes the installation procedure for installing {product} on your OpenShift Container Platform using the Operator.

//You can use the {product} Operator in Red Hat OpenShift (OpenShift) to install {product}.

.Prerequisites

* You are logged in as an administrator on the OpenShift web console.
* You have configured the appropriate roles and permissions within your project to create an application. See the link:https://docs.openshift.com/container-platform/4.14/applications/index.html[Red Hat OpenShift documentation on Building applications] for more details.
* If installing a CI build from quay.io/rhdh, make sure that:
** you have been granted permission to pull images from the `rhdh` organization on quay.io, and 
** you have added your Quay token to the cluster global pull secret (link:https://docs.openshift.com/container-platform/4.14/openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets[link]):

> {code}

Reviewer 3: ```suggestion
```
```

> {code}

Reviewer 3: ```suggestion
[source]
----
```

oc get secret/pull-secret -n openshift-config --template='{{index .data ".dockerconfigjson" | base64decode}}' > /tmp/my-global-pull-secret.yaml

oc registry login --registry="quay.io" --auth-basic="<user>:<token>" --to=/tmp/my-global-pull-secret.yaml

oc set data secret/pull-secret -n openshift-config --from-file=.dockerconfigjson=/tmp/my-global-pull-secret.yaml

rm -f /tmp/my-global-pull-secret.yaml

> {code}

Reviewer 3: ```suggestion
----
```

> ** you have added a link:https://gist.github.com/nickboldt/d54ba9d7264a5480d14eea6c3bd2ecdf[RHDH Operator Catalog Source]:

Reviewer 3: This is OK for now but we should move this script into a public place like either THIS repo or the future home of https://github.com/janus-idp/backstage-showcase/ so it's available in the docs. 

Should be fixed before GA... if we can move https://github.com/janus-idp/backstage-showcase/ before 1.1, then use that; if not, then embed a copy of the script in this repo.

Reviewer: > or the future home of [janus-idp/backstage-showcase](https://github.com/janus-idp/backstage-showcase/)

You mean "the future home of [janus-idp/operator](https://github.com/janus-idp/operator)", right?

Reviewer 3: yes, installer script makes more sense in the operator repo than the hub repo, as we can also install hub via helm.  Sorry, wrong link paste.

> {code}

Reviewer 3: ```suggestion
[source]
----
```

curl -L https://gist.githubusercontent.com/nickboldt/d54ba9d7264a5480d14eea6c3bd2ecdf/raw/76a7f166a17f86ae8e766241b9d648c9821898bb/install-rhdh-next-build.sh | bash

> {code}

Reviewer 3: ```suggestion
----
```

.Procedure

> . In the *Administrator* view of the OpenShift web console, go to *Operators* → *OperatorHub*, search for Red Hat Developer Hub, and install the Red Hat Developer Hub Operator. For more info, see link:https://docs.openshift.com/container-platform/4.14/operators/admin/olm-adding-operators-to-cluster.html#olm-installing-from-operatorhub-using-web-console_olm-adding-operators-to-a-cluster[Installing from OperatorHub using the web console].

Reviewer 3: ```suggestion
. In the *Administrator* perspective of the OpenShift web console, go to *Operators* → *OperatorHub*, search for Red Hat Developer Hub, and install the Red Hat Developer Hub Operator. For more info, see link:https://docs.openshift.com/container-platform/4.14/operators/admin/olm-adding-operators-to-cluster.html#olm-installing-from-operatorhub-using-web-console_olm-adding-operators-to-a-cluster[Installing from OperatorHub using the web console].
```

. Create an OpenShift project to be used by your Backstage instance. For more information about creating a project in OpenShift, see the link:https://docs.openshift.com/container-platform/4.14/applications/projects/working-with-projects.html#creating-a-project-using-the-web-console_projects[Red Hat OpenShift documentation].

> . Switch to *Developer* mode on your Red Hat OpenShift web console.

Reviewer 3: ```suggestion
. Administrative and non-administrative users can now use the *Developer* perspective in the Red Hat OpenShift web console. Switch to that perspective now.
```
In the Getting Started Guide (GSG) we should start the install instructions here, with a prereq that your administrator has already installed the operator subscription, with a link to that section of doc in the AG (Admin Guide).
